### DISCUSSION TITLE
platform_keyboard_set_layout for a french layout.

### DISCUSSION THREAD

**Discussion by mvvvv** [StereoKit/StereoKit#886] [Q&A]
Hello, I created the beginning of a layout for a French keyboard. French keyboards (and other languages) use the alt key for at least 13 characters including @ { } [ ] | . Would it be possible to implement a layout for the Alt key?
![French_Maj](https://github.com/StereoKit/StereoKit/assets/810124/baa999c7-ebdc-44dc-8a37-f3fc25e0253a)
![French_minuscules](https://github.com/StereoKit/StereoKit/assets/810124/4725ad29-530b-447f-883f-e82eb6d301ab)

The code is there: https://github.com/mvvvv/StereoKit-rust/blob/master/examples/demos/text1.rs and it's the third layout: FR_KEY_TEXT_ALT  I would like to link to the Alt key.

I ran into some small issues with UI.Input but it might be a little early to bother you with that and you should run into them soon. Let me know.

**Comment by maluoi** [comment]
You can probably do this using `visit_2` instead of `mod`, at least in the short term! This will generate the correct text characters, but might not send quite the right keyboard keycodes for the `alt` key.

I'm uncertain if sending the right keystrokes is critical with the soft keyboard, especially since it's primarily about text entry and not keyboard control of the app? I'm interested in feedback on that, but I'm currently planning on removing most of the VK codes once `UI.Input` gets touch controls. So keyboard layouts might look a lot more like this in the future:
```
layout =
R"(q|w|e|r|t|y|u|i|o|p
---1|a|s|d|f|g|h|j|k|l
spr:sk/ui/shift---3-go_1|z|x|c|v|b|n|m|spr:sk/ui/backspace-\b--3
spr:sk/ui/close----close|123---3-go_2|,| - --7|.|Return-\n--4|)";
```

And yeah, I've _probably_ seen what you're talking about with input elements. Selection and the cursor there need some attention 😅

**Reply by mvvvv** [reply]
Thanks, Ill try that tomorrow and will tell you.

**Reply by mvvvv** [reply]
I published the French virtual keyboard with 3 layouts and using -go_ rather than -visit_ on SHIFT and ALT.
The code is there: https://github.com/mvvvv/StereoKit-rust/blob/master/examples/demos/text1.rs

With 3 layouts it does not seem possible to combine -go_ and -visit_ and I therefore chose -go_ to be able to easily write an entire text in capitals. It's still a little confusing because with the ALT layout I spend long seconds trying to understand what I have in front of me. Would it be possible to add a single color per layout?
Or otherwise, to avoid the 3rd layout, there is also the solution used on smartphones and in quests, to bring up a submenu when holding down certain keys. Is this possible here?
Would it be possible to have functional "Del" and "insert" keys?

Video  : https://youtu.be/hUp_-dwzE5w

Some problems encountered:
* At the end of the video I made a few moves to show glitches and the display of text at the back of the panel (beyond 2 or 3 meters). Thanks for fixing the waltz when rotating the camera by the way.
* Password type UI.Input crash the application when selected on Quest.
* UI.Input:
   - Must be non-empty initially then left arrow to be able to insert text.
  - Would deserve to have the attributes of UI.Text.
  - up and down arrows have no effect in multi-line.
  - Possibility to select an area of the text to modify
  - deserve the emoji (just like the keyboard)
  - deserve Copy/Paste

Sorry, we already have a lot but we always want more.
